# Introduction
In the 2nd component of the project, I will be using `pandas` in Jupyter Notebook to demonstrate the data exploration techniques that I will usually apply to the data set. As I do have some background knowledge of CS GO Skin, I have made some specific searches to find out more about the data. You could refer to this [link](https://counterstrike.fandom.com/wiki/Skins) for more info.

The data extracted from the 1st component was saved as a csv file - [Date of Extraction]_Market_List.

In [1]:
# import the pandas library and fetch the data
import pandas as pd
df = pd.read_csv("../data/item_list/20230619_2321_v3.csv") # change to your own file

# Generic Data Exploration

In [2]:
# find out the number of rows and columns of the dataframe
df.shape

(20457, 9)

In [3]:
'''
find out non-null values in the dataframe, which indirectly allows us to check on 
presence of null values when compared to the row value from `df.shape`
'''
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20457 entries, 0 to 20456
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   data_start_value  20457 non-null  int64 
 1   classid           20457 non-null  int64 
 2   instanceid        20457 non-null  int64 
 3   tradable          20457 non-null  int64 
 4   skin_type         20457 non-null  object
 5   market_name       20457 non-null  object
 6   market_hash_name  20457 non-null  object
 7   commodity         20457 non-null  int64 
 8   timestamp         20457 non-null  object
dtypes: int64(5), object(4)
memory usage: 1.4+ MB


In [4]:
# peak at the first 5 rows of data
df.head()

data_start_value     classid  instanceid  tradable          skin_type  \
0                 0  5381859093   519977179         1  Exceptional Agent   
1                 1  4587946730   519977179         1       Master Agent   
2                 2  3729791865   519977179         1       Master Agent   
3                 3  5380326558   519977179         1     Superior Agent   
4                 4  3608084078   519977179         1     Superior Agent   

                                    market_name  \
0            'Blueberries' Buckshot | NSWC SEAL   
1  'Medium Rare' Crasswater | Guerrilla Warfare   
2                  'The Doctor' Romanov | Sabre   
3              'Two Times' McCoy | TACP Cavalry   
4                 'Two Times' McCoy | USAF TACP   

                               market_hash_name  commodity  \
0            'Blueberries' Buckshot | NSWC SEAL          0   
1  'Medium Rare' Crasswater | Guerrilla Warfare          0   
2                  'The Doctor' Romanov | Sabre          0   
3              'Two Times' McCoy | TACP Cavalry          0   
4                 'Two Times' McCoy | USAF TACP          0   

                    timestamp  
0  2023-06-19 23:21:24.162178  
1  2023-06-19 23:21:24.162178  
2  2023-06-19 23:21:24.162178  
3  2023-06-19 23:21:24.162178  
4  2023-06-19 23:21:24.162178

In [5]:
# quick way to get all the descriptive analysis of columns
df.describe(include="all")

data_start_value       classid    instanceid  tradable  \
count       20457.000000  2.045700e+04  2.045700e+04   20457.0   
unique               NaN           NaN           NaN       NaN   
top                  NaN           NaN           NaN       NaN   
freq                 NaN           NaN           NaN       NaN   
mean        10228.000000  3.980993e+09  2.543854e+08       1.0   
std          5905.571564  1.722328e+09  1.552556e+08       0.0   
min             0.000000  1.861506e+08  0.000000e+00       1.0   
25%          5114.000000  2.660190e+09  1.885301e+08       1.0   
50%         10228.000000  5.051655e+09  1.885301e+08       1.0   
75%         15342.000000  5.380619e+09  1.885349e+08       1.0   
max         20456.000000  5.382769e+09  1.597846e+09       1.0   

                 skin_type                         market_name  \
count                20457                               20457   
unique                 115                               20457   
top     High Grade Sticker  'Blueberries' Buckshot | NSWC SEAL   
freq                  2183                                   1   
mean                   NaN                                 NaN   
std                    NaN                                 NaN   
min                    NaN                                 NaN   
25%                    NaN                                 NaN   
50%                    NaN                                 NaN   
75%                    NaN                                 NaN   
max                    NaN                                 NaN   

                          market_hash_name     commodity  \
count                                20457  20457.000000   
unique                               20457           NaN   
top     'Blueberries' Buckshot | NSWC SEAL           NaN   
freq                                     1           NaN   
mean                                   NaN      0.411546   
std                                    NaN      0.492126   
min                                    NaN      0.000000   
25%                                    NaN      0.000000   
50%                                    NaN      0.000000   
75%                                    NaN      1.000000   
max                                    NaN      1.000000   

                         timestamp  
count                        20457  
unique                         268  
top     2023-06-19 23:21:24.162178  
freq                           100  
mean                           NaN  
std                            NaN  
min                            NaN  
25%                            NaN  
50%                            NaN  
75%                            NaN  
max                            NaN

# Drill-down

While the `describe` method allows for quick analysis, they selectively analyse the columns based on the datatype. For instance, they won't analyse the unique count of int/float columns. So, further steps are required for this.

In [6]:
# check the unique values of each column
col_unique_count = pd.DataFrame(df.apply(lambda col: len(col.unique())))
col_unique_count.reset_index().rename({"index":"columns", 0:"unique_count"}, axis="columns")

columns  unique_count
0  data_start_value         20457
1           classid         20456
2        instanceid           369
3          tradable             1
4         skin_type           115
5       market_name         20457
6  market_hash_name         20457
7         commodity             2
8         timestamp           268

Considering that there was only 1 duplicated row as compared to the total rows, investigated the "classid" column through `df[df["classid"].duplicated(False)]` to see all the duplicated rows. From what I can see, it is basically a duplication of name, as seen in the lack of spacing between "Ivette" and "(Holo)" in data index 13763.

In [7]:
# examine the 1 duplicated value
df[df["classid"].duplicated(False)]

data_start_value     classid  instanceid  tradable           skin_type  \
13762             13762  1440563571   188530139         1  Remarkable Sticker   
13763             13763  1440563571   188530139         1  Remarkable Sticker   

                   market_name         market_hash_name  commodity  \
13762  Sticker | Ivette (Holo)  Sticker | Ivette (Holo)          1   
13763   Sticker | Ivette(Holo)   Sticker | Ivette(Holo)          1   

                        timestamp  
13762  2023-06-20 00:05:00.892388  
13763  2023-06-20 00:05:00.892388

Proceeded to check how "instanceid" column works but looking at how the data is skewed towards the top few instances. While it may be interesting to investigate, it does not add value to the objective of this project. It could be used for future investigations, finding out what this column meant.

In [8]:
# check how "instanceid" categorizes the data
df.groupby(['instanceid']).count().sort_values("data_start_value", ascending=False)

data_start_value  classid  tradable  skin_type  market_name  \
instanceid                                                                
188530139              11124    11124     11124      11124        11124   
188530170               2512     2512      2512       2512         2512   
519977179               2151     2151      2151       2151         2151   
302028390               1410     1410      1410       1410         1410   
480085569                755      755       755        755          755   
...                      ...      ...       ...        ...          ...   
188642649                  1        1         1          1            1   
188651524                  1        1         1          1            1   
188661951                  1        1         1          1            1   
188662885                  1        1         1          1            1   
1597846221                 1        1         1          1            1   

            market_hash_name  commodity  timestamp  
instanceid                                          
188530139              11124      11124      11124  
188530170               2512       2512       2512  
519977179               2151       2151       2151  
302028390               1410       1410       1410  
480085569                755        755        755  
...                      ...        ...        ...  
188642649                  1          1          1  
188651524                  1          1          1  
188661951                  1          1          1  
188662885                  1          1          1  
1597846221                 1          1          1  

[369 rows x 8 columns]

Next, we see that tradable has only 1 as the unique value. Judging from how conventionally people store data of "-able" columns as Binary, meaning 1 as positive and 0 as negative, having 1 in this column means the item is tradable. This allows us to deduce that all items that will appear in any future extractions are tradable.

In [9]:
df["tradable"].unique()

array([1], dtype=int64)

"commodity" column is also binary, and with further investigation, we can see that skins that are considered commodity are non-weapon and non-agent.

In [10]:
df['commodity'].unique()

array([0, 1], dtype=int64)

In [11]:
df[df['commodity']==1]["skin_type"].unique()

array(['Base Grade Container', 'Exotic Collectible',
       'Extraordinary Collectible', 'Base Grade Pass', 'Base Grade Gift',
       'High Grade Collectible', 'Base Grade Key',
       'Remarkable Collectible', 'High Grade Music Kit', 'Base Grade Tag',
       'High Grade Patch', 'Exotic Patch', 'Remarkable Patch',
       'High Grade Graffiti', 'Base Grade Graffiti', 'Exotic Graffiti',
       'Remarkable Graffiti', 'StatTrak™ High Grade Music Kit',
       'Base Grade Tool', 'Remarkable Sticker', 'Extraordinary Sticker',
       'Exotic Sticker', 'High Grade Sticker', 'Contraband Sticker'],
      dtype=object)

We then move on to check on more specific details within the skin_type column as I wanted to check on some of the weapons. From this we can see that "AK-47 | Aquamarine Revenge" has a complete set of 5 skin wear.

In [12]:
df[df["skin_type"].str.contains("Rifle")].head()

data_start_value     classid  instanceid  tradable     skin_type  \
13                13  5381883863   188530139         1  Covert Rifle   
14                14  5380621322   480085569         1  Covert Rifle   
15                15  1011938131   480085569         1  Covert Rifle   
16                16  5374387479   188530139         1  Covert Rifle   
17                17  1011942929   188530139         1  Covert Rifle   

                                    market_name  \
13  AK-47 | Aquamarine Revenge (Battle-Scarred)   
14     AK-47 | Aquamarine Revenge (Factory New)   
15    AK-47 | Aquamarine Revenge (Field-Tested)   
16    AK-47 | Aquamarine Revenge (Minimal Wear)   
17       AK-47 | Aquamarine Revenge (Well-Worn)   

                               market_hash_name  commodity  \
13  AK-47 | Aquamarine Revenge (Battle-Scarred)          0   
14     AK-47 | Aquamarine Revenge (Factory New)          0   
15    AK-47 | Aquamarine Revenge (Field-Tested)          0   
16    AK-47 | Aquamarine Revenge (Minimal Wear)          0   
17       AK-47 | Aquamarine Revenge (Well-Worn)          0   

                     timestamp  
13  2023-06-19 23:21:24.162178  
14  2023-06-19 23:21:24.162178  
15  2023-06-19 23:21:24.162178  
16  2023-06-19 23:21:24.162178  
17  2023-06-19 23:21:24.162178

As I explored further along the column by using the `iloc` method, I found that not all weapons have all 5 skin wear. This can be seen in the "AK-47 | X-Ray" skin, where "Minimal Wear" and "Factory New" skins are missing. This means that the items lists might not be full and requires further addition.

In [13]:
df[df["skin_type"].str.contains("Rifle")].iloc[200:210]

data_start_value     classid  instanceid  tradable             skin_type  \
213               213   638244432   480085569         1          Covert Rifle   
214               214  5109396709   188530139         1          Covert Rifle   
215               215  4145213923   519977179         1          Covert Rifle   
216               216  4472567108   519977179         1          Covert Rifle   
217               217  4141827845   519977179         1          Covert Rifle   
248               248  5347667128   519977179         1          Covert Rifle   
249               249  5381065584   302028390         1          Covert Rifle   
250               250  5188286021   302028390         1          Covert Rifle   
251               251  5376857350   519977179         1  Mil-Spec Grade Rifle   
252               252  4578724327   519977179         1  Mil-Spec Grade Rifle   

                                 market_name  \
213   AK-47 | Wasteland Rebel (Minimal Wear)   
214      AK-47 | Wasteland Rebel (Well-Worn)   
215           AK-47 | X-Ray (Battle-Scarred)   
216             AK-47 | X-Ray (Field-Tested)   
217                AK-47 | X-Ray (Well-Worn)   
248  AUG | Akihabara Accept (Battle-Scarred)   
249    AUG | Akihabara Accept (Field-Tested)   
250       AUG | Akihabara Accept (Well-Worn)   
251           AUG | Amber Fade (Factory New)   
252          AUG | Amber Fade (Field-Tested)   

                            market_hash_name  commodity  \
213   AK-47 | Wasteland Rebel (Minimal Wear)          0   
214      AK-47 | Wasteland Rebel (Well-Worn)          0   
215           AK-47 | X-Ray (Battle-Scarred)          0   
216             AK-47 | X-Ray (Field-Tested)          0   
217                AK-47 | X-Ray (Well-Worn)          0   
248  AUG | Akihabara Accept (Battle-Scarred)          0   
249    AUG | Akihabara Accept (Field-Tested)          0   
250       AUG | Akihabara Accept (Well-Worn)          0   
251           AUG | Amber Fade (Factory New)          0   
252          AUG | Amber Fade (Field-Tested)          0   

                      timestamp  
213  2023-06-19 23:21:39.127844  
214  2023-06-19 23:21:39.127844  
215  2023-06-19 23:21:39.127844  
216  2023-06-19 23:21:39.127844  
217  2023-06-19 23:21:39.127844  
248  2023-06-19 23:21:39.127844  
249  2023-06-19 23:21:39.127844  
250  2023-06-19 23:21:39.127844  
251  2023-06-19 23:21:39.127844  
252  2023-06-19 23:21:39.127844

It then occurred to me to check a skin that used to be prized back in 2012/13 when I was still an active "trader". I searched specifically for a rifle with the name Dragon Lore. While the skin name is "AWP | Dragon Lore", if I were to search with the code `df[df["market_hash_name"].str.contains("AWP | Dragon Lore")]`, it will not work. This is because this search method will interpret `|` as the OR command, thus search for string that contains "AWP" or "Dragon Lore". Hence, the below code was used. 

As the search results returned empty, we can say that this item was not extracted. After some research online, I found that this API only returns items that are currently on the market. So the data cleaning needs to take into account this point.

In [14]:
df[(df["market_hash_name"].str.contains("Dragon Lore")) & (df["market_hash_name"].str.contains("AWP"))]

Empty DataFrame
Columns: [data_start_value, classid, instanceid, tradable, skin_type, market_name, market_hash_name, commodity, timestamp]
Index: []

# Conclusion/Findings

1. There are no null values in the data-set of 20457 rows.
2. There is 1 duplicated "classid" value, and the items with the same class id has similar skin name with only 1 space difference. This suggests a data-error from the source, which should be removed during analysis.
3. "instanceid" will not be utilized for this current project.
4. There are 2 limitation to the data due to the fact that skins are only present in the extraction if they are currently on the market:
    - Skins having less than 5 skin wear at point of extraction
    - Missing skins from the extraction